### PR TITLE
Adding and implementing interfaces to make possible to add extension attributes to cms page and block.

### LINKED ISSUE #17458: CMS module does not implement extension attributes
Currently the entities provided by the CMS module do not allow the use of extension attributes. Is there a reason for this? Or is it on the roadmap to add this functionality?

### Preconditions

1. Magento 2.x

### Steps to reproduce

1. Try to add an extension attribute to `Magento\Cms\Model\Page`.

### Expected result

1. Extension attributes should be added / registered to the model.

### Actual result

1. They're not

### PR DESCRIPTION
**Pull request by bartoszkubicki**
### Description
Until now, there was no possibility to extend page and block models using extension attributes. Model has been fixed to enabled this option. According to backward compatibility models implements new interface, no changes were made in existing ones.

### Fixed Issues (if relevant)

1. Fixes https://github.com/magento/magento2/issues/17458

### Manual testing scenarios
1. Add custom extension attribute 
2. Try if you can save and get it.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=17995) <br/>All committers have signed the CLA.

**Comment by magento-engcom-team** [comment]
Hi @bartoszkubicki. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me {$VERSION} instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Comment by ishakhsuvarov** [comment]
Hi @bartoszkubicki 

This fix seems to solve problem only partially, as the original contract still does not have extension attributes, therefore, any clients would still be unable to use it correctly. In this specific case I would suggest deprecating the old interface, and creating a new one, which extends original but adds extension attributes to it. What do you think on this?

**Comment by bartoszkubicki** [comment]
Hi @ishakhsuvarov,
actually I was advised to do it that way in order to support backward compatibility, but in fact is seems to be not supported. 
Well, I am going to modify it, I will mark old interface as deprecated, but I have to questions:
- doing it that way maybe naming on new interface should be different (as it it would be new service contract for block and pages) - maybe BlockExtensibleInterface?
- if we deprecate it, it is going to be removed completely some day - so is inheritance way to go in this case? If we later remove old interface whole service contract will break down. Copying all methods definitions and not using inheritance maybe is an option, if we want to prevent any future problem. Maybe some more clever way?

**Comment by bartoszkubicki** [comment]
@ishakhsuvarov I remind about my questions.

**Comment by ishakhsuvarov** [comment]
@bartoszkubicki 
* `BlockExtensibleInterface` or `ExtensibleBlockInterface` seems reasonable to me.
* I think that copying methods and proxying implementation appropriately is a way to go in this case. There is some more information on the subject in the [Backward compatible development](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/) documentation (Section `Introduction of a method to a class or interface`)

**Comment by bartoszkubicki** [comment]
@ishakhsuvarov Thanks for responses, but I still have other doubts. Ok, I see that we are not able to modify BlockInterface/PageInterface according to backward compatibility issues. But, does new interface solve any issue? Ok it will exist, but if developer still won't be able to extend entities with existing attributes, as in all places still these deprecated ones will be used. Or - I have missed something? Should I add something more, to have it work, right out of the box?

**Comment by sivaschenko** [comment]
Hi @bartoszkubicki , my understanding is: developers will be able to use the new interface for new implementations. As old interfaces will be marked as deprecated, we will be able to remove them and refactor the usages later on.

**Comment by bartoszkubicki** [comment]
Ok, understood, in a few days I will provide some changes according to guidelines.

**Comment by sidolov** [comment]
Hi @bartoszkubicki , will you continue progress with the PR?

**Comment by bartoszkubicki** [comment]
Yes, I didn't have time in janury. I will finish it until the end of january.

**Comment by bartoszkubicki** [comment]
@ishakhsuvarov @sivaschenko @sidolov 

I have started working on this PR again, taking into account your notes. 
It is not done yet, I need some help:
- I am creating new interfaces - so I decided to change some types returned by some functions - I limited in many funcitons returning null or string to string only. Is it correct?
- I have removed some deprecated functionalities - is it ok?
- how can fullfill sentence from backward compatibility guide: "old methods should proxy the calls to the new interface instead of duplicating the logic." - can you explain it on some examples from code provided
- generally - is it now heading to right direction?

TODO:
- proxying between implementations
- unit tests for new implementations

**Comment by sivaschenko** [comment]
Hi @bartoszkubicki 

- I think it's acceptable to specify the correct return types in new interfaces
- Not including deprecated functionality within new implementation seems ok to me
- I don't think "old methods should proxy the calls to the new interface instead of duplicating the logic." sentence is applicable to the implementation in this pull request, as the implementation provided introduces new entities without replacing old ones

As for confirmation that the direciton is right, I would participate at the public architectural discussion meeting, introduce the proposal and get feedback from magento architects.

You can present your proposal at the public architectural discussion meetitng by:
 - adding the pull request to agenda - as a comment to corresponding issue here: https://github.com/magento/architecture/issues
 - joining the actual meeting using the following information https://github.com/magento/architecture/wiki#architectural-discussion-meetings

**Comment by Vinai** [comment]
Just because I often provide negative feedback, here is something different:
I would like to take the opportunity to state I agree with you. Creating a new interface is a good way to go for this case!
Also, thank you for preserving backward compatibility!
Great work!

**Comment by bartoszkubicki** [comment]
@sivaschenko Any updates? Any things to do it in interfaces or implementations or should I proceed to unit tests?

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Hi @bartoszkubicki we've performed a group code review of this pull request together with maintainers and have a couple of concerns:

There is no clear way on how new interfaces/implementation can be integrated in the system. To make them compatible with the current code base:
 - the new interfaces have to extend the actual interfaces (introducing only the methods for extension attributes handling)
 - the DI preference for existing Block/Page Interfaces should be updated to use the new models (while keeping the existing preference)

I think it will be necessary for new models to extend the existing models in order to keep support of third-party plugins.
 
It is still a good idea to discuss this pull request at the architecture discussion meeting as I mentioned in the previous comment. Please let me know if you need any help to organize this.

**Comment by bartoszkubicki** [comment]
@sivaschenko Yes, I was wondering a few comments ago, how later incorporate this into existing code and replace usage of old implementation.

I think for new I understand what have to be done and I will try to acomplish this task in a few days.

**Comment by bartoszkubicki** [comment]
@sivaschenko points 1 and 2 are done, although I have questions:
- as we stepped back with interfaces a little I have to change implementations, so they are compatible with old interface (I mean return types, which I've added) - so along inheritance in interfaces should I come back to inheritance in models also - So `BlockExtensible` inherits from `Block`?
- I understand we leave deprecated notes in all places where I've already put it? (old interfaces and models?)

**Comment by sivaschenko** [comment]
@bartoszkubicki considering that new interfaces will extend the existing ones, methods containing in the existing interfaces should not be deprecated (as they will be still used).

As for the model, if you will go without extending the existing model (that can make it a bit harder to integrate new models, but simplifies the complete transition to new models and ability to remove the old models) - the old models and methods should be marked as deprecated

**Comment by bartoszkubicki** [comment]
@sivaschenko I have pushed changes according to your guidelines. Please make sure all is correct and I will write missing unit tests so we can proceed to testing whole PR.

**Comment by sivaschenko** [comment]
Hi @bartoszkubicki thanks for the updates, the PR looks fine, the only thing I'd suggest is to mark existing models as deprecated.

Can you also take a look at the failed tests?

**Comment by bartoszkubicki** [comment]
@sivaschenko Sure, I will do it in a couple of days.

**Comment by bartoszkubicki** [comment]
@sivaschenko Interface for tests in github must have changed.. When I click on tests above I can only read that there was failure, but no more deatils. Is it available on github? I don't have each test type configured locally, and some mistakes could be easily fixed without setting up tests locally (like some formatting issues)

**Comment by sivaschenko** [comment]
@bartoszkubicki the tests are failing with the message "Please merge the latest mainline to you Pull Request". Looks like there are some issues with updating the pull request from mainline automatically.

Can you please try to merge the latest mainline manually and see if it will fix the tests

**Comment by bartoszkubicki** [comment]
@sivaschenko What does it mean latest mainline? It means that thera git conflicts or somethin?

**Comment by sivaschenko** [comment]
@bartoszkubicki I propose to simply try to update your fork/branch from magento/magento2

**Comment by bartoszkubicki** [comment]
@sivaschenko I'll do it soon, but first I have pushed unit tests from previous models with some adjustments.

Some more doubts:

- I have added some TODOs in code. Please review them, they are mainly about abonding some functions in new implementations. 
- If I have function in new implementation returning $this - return type should be: `self` or `PageInterface` or `PageExtensibleInterface`?

File: app/code/Magento/Cms/Model/BlockExtensible.php [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+use Magento\Cms\Model\ResourceModel\Block as BlockResource;
+use Magento\Framework\DataObject\IdentityInterface;
+use Magento\Framework\Exception\LocalizedException;
+use Magento\Framework\Model\AbstractExtensibleModel;
+
+/**
+ * Class BlockExtensible
+ * @package Magento\Cms\Model
+ *
+ * TODO: Maybe this should be part of interface and we get rid of magic here
```
Comment (sivaschenko) [inline app/code/Magento/Cms/Model/BlockExtensible.php:24]: I think these methods should be added to interface

File: app/code/Magento/Cms/Model/BlockExtensible.php [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+        } elseif (!empty($storeDataByStoreIdKey)) {
+            return is_array($storeDataByStoreIdKey) ? $storeDataByStoreIdKey : [$storeDataByStoreIdKey];
+        }
+
+        return [];
+    }
+
+    /**
+     * TODO: Should this be part of interface? Using public function not in available in interface makes us depend on
+     * TODO: concrete implementation - maybe this should be marked as depracted and moved to separte service contract?
```
Comment (sivaschenko) [inline app/code/Magento/Cms/Model/BlockExtensible.php]: I think this method should be marked as deprecated and moved to an instance of OptionSourceInterface

File: app/code/Magento/Cms/Model/BlockExtensible.php [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+        }
+
+        throw new LocalizedException(
+            __('Make sure that static block content does not reference the block itself.')
+        );
+    }
+
+    /**
+     * TODO: Should this be part of interface? Or maybe we should introduce new functions
+     * TODO: to get storeIds and depracate this one?
```
Comment (sivaschenko) [inline app/code/Magento/Cms/Model/BlockExtensible.php]: I think this method should be added to interface

File: app/code/Magento/Cms/Model/PageExtensible.php [hunk 4]
```diff
@@ -0,0 +1,583 @@
...
+                    __('This identifier is reserved for "CMS No Cookies Page" in configuration.')
+                );
+        }
+
+        return parent::beforeSave();
+    }
+
+
+    /**
+     * TODO: Maybe it could be depracted, if other model functions are deprecated?
```
Comment (sivaschenko) [inline app/code/Magento/Cms/Model/PageExtensible.php]: yes, it should be deprecated in favour to \Magento\Cms\Model\ResourceModel\Block::load

File: app/code/Magento/Cms/Model/PageExtensible.php [hunk 5]
```diff
@@ -0,0 +1,583 @@
...
+        if ($id === null) {
+            return $this->noRoutePage();
+        }
+
+        $this->_resource->load($this, $id, $field);
+        return $this;
+    }
+
+    /**
+     * TODO: Should this be part of interface?
```
Comment (sivaschenko) [inline app/code/Magento/Cms/Model/PageExtensible.php]: I think this method should be marked as deprecated and moved to a "Management" class

File: app/code/Magento/Cms/Model/PageExtensible.php [hunk 6]
```diff
@@ -0,0 +1,583 @@
...
+     * @return self
+     */
+    public function noRoutePage(): self
+    {
+        return $this->load(self::NOROUTE_PAGE_ID, $this->getIdFieldName());
+    }
+
+    /**
+     * TODO: Should this be part of interface? Using public function not in available in interface makes us depend on
+     * TODO: concrete implementation - maybe this should be marked as depracted and moved to separte service contract?
```
Comment (sivaschenko) [inline app/code/Magento/Cms/Model/PageExtensible.php]: I think this method should be marked as deprecated, `\Magento\Cms\Api\GetPageByIdentifierInterface::execute` can be used to check if the page exists

File: app/code/Magento/Cms/Model/PageExtensible.php [hunk 7]
```diff
@@ -0,0 +1,583 @@
...
+     * @throws LocalizedException
+     */
+    public function checkIdentifier(string $identifier, int $storeId): int
+    {
+        return (int) $this->_resource->checkIdentifier($identifier, $storeId);
+    }
+
+    /**
+     * TODO: Should this be part of interface? Using public function not in available in interface makes us depend on
+     * TODO: concrete implementation - maybe this should be marked as depracted and moved to separte service contract?
```
Comment (sivaschenko) [inline app/code/Magento/Cms/Model/PageExtensible.php]: I think this method should be marked as deprecated and moved to a separate class

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Hi @bartoszkubicki please see my responses to TODO items.

As for return type annotation, I think `@return $this` should be fine

**Comment by bartoszkubicki** [comment]
@sivaschenko Ok I will try to resolve all discussions according to guidelines. 
-> If I have function in new implementation returning $this - return type should be: self or PageInterface or PageExtensibleInterface? - I meant not annotations but real return types - what types should be used accross new Model?

**Comment by sivaschenko** [comment]
@bartoszkubicki I didn't understand your question. What do you mean by real return type?

**Comment by bartoszkubicki** [comment]
@sivaschenko I mean just php return type in function declaration, not in annotations

**Comment by sivaschenko** [comment]
@bartoszkubicki it should be `PageExtensibleInterface` for functions that are supposed to return self and are declared in the `PageExtensibleInterface` (and the model implementing this interface correspondingly)

**Comment by bartoszkubicki** [comment]
@sivaschenko Branch is updated I think, we will see now if tests are green, then I will complete all TODO and suggests.

**Comment by bartoszkubicki** [comment]
@sivaschenko ok, some tests are still failing. Am I allowed in any way to see details? If I click 'Details' I see nothing more than here. Still branch is not up to date?

**Comment by sivaschenko** [comment]
@bartoszkubicki test reports can be accessed via the following way:
1. Click details next to failed test (or open Checks tab of the pull request)
2. Select the failed test
3. Click on the report link in the "Build reports" section

image

File: app/code/Magento/Cms/Model/BlockExtensible.php (hunk 1 above)
Comment (bartoszkubicki) [reply app/code/Magento/Cms/Model/BlockExtensible.php:24]: On the other hand we have getStores method widely used - maybe we should choose just one? 

For example getStores(): array and setStores(array $stores)?

File: app/code/Magento/Cms/Model/PageExtensible.php (hunk 5 above)
Comment (bartoszkubicki) [reply app/code/Magento/Cms/Model/PageExtensible.php]: do you mean some concrete existing implementation or it requires creating this class?

**Comment by bartoszkubicki** [comment]
@sivaschenko I have resolved most of TODOs, I have left you some more questions about TODOs, and also fixed a little return types and unit tests a little.

I have also reviewed tests fails. Static tests says that `PageExtensible` has too much methods, but we have decided to copy all methods to make removing `Page` more easily - what can we do here? Surppres warning? I am not sure if I am able to fix it, as it may be some legacy properties. I am a little confused about integration tests and functional tests and magento index health. Failures seems to be not connected with changes I've made. Any help with these ones especially would be welcome.

File: app/code/Magento/Cms/Model/BlockExtensible.php (hunk 1 above)
Comment (sivaschenko) [reply app/code/Magento/Cms/Model/BlockExtensible.php:24]: I'd prefer to keep only getStores/setStores methods in the interface, however as the new interfaces are extending and replacing the existing interfaces with the setStoreId/getStoreId methods mentioned, I'd keep these methods for the compatibility purposes with the implementation proxying to setStores/getStores methods

File: app/code/Magento/Cms/Model/PageExtensible.php (hunk 5 above)
Comment (sivaschenko) [reply app/code/Magento/Cms/Model/PageExtensible.php]: It's better to create a new class

Comment (sivaschenko) [reply app/code/Magento/Cms/Model/PageExtensible.php]: I think it's better to create a new class

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Hi @bartoszkubicki thanks for the updates! Please see my responses in the review

**Comment by sivaschenko** [comment]
@magento run all tests

**Comment by sivaschenko** [comment]
Test results look weird, relaunched tests to confirm

**Comment by bartoszkubicki** [comment]
@sivaschenko Currently (I think temporarily) I lack free time to continue working on this PR. If this feature is very urgent, feel free to share a work with someone. BTW, any clue about weird tests results?

**Comment by sivaschenko** [comment]
@magento run all tests

**Comment by sivaschenko** [comment]
@bartoszkubicki I think that test results are correct and related to the changes

**Comment by bartoszkubicki** [comment]
@sivaschenko Ok. I hope next month I will be able to do some work.

**Comment by sidolov** [comment]
Hi @bartoszkubicki will you continue progress?

**Comment by bartoszkubicki** [comment]
yes, I have some not finished, uncomitted changes, I will post it at saturday

**Comment by bartoszkubicki** [comment]
Ok, I have pushed some significant fixes, and I have seen (before last TODO update) all major tests passed. Only some static checks are still pending, but can be fixed by decisions which are to make with some TODOs included in code. Plase review these notes and comment, what should be done. I feel some methods could be moved to separate service contracts or refactored. Please let me know if this action are in the scope of this PR.

I have reordered methods in models, so they are in the same sort order as in previous implementations - it is for easier comparing of code. Later when service contract is finished I bet they will dissappear.

Apart of these more serious things, I have to fix getters in both models to return always correct return types (we can set using `setData()` what we want as model value) + fix missing unit tests.
Also you may note, that I have edited both `Repository` classes. It was due to failing web api functional tests. It turned out that there are mutliples places where there is `Magento\Cms\Model\BlockFactory` or `PageFactory` injected. In `Repositories` there were also `BlockInterfaceFactory` and `PageInterfaceFactory` already injected, so I just could replace dependency usage in methods, but we have no such a luck in all places where factories are present. I need advice, what we do about that, as type inconsistency may cause some errors. Of course, backward compatability must be kept - but the question is if further changes in other classes are in scope of this PR.

**Comment by bartoszkubicki** [comment]
Sorry guys, while working on on this PR, I have messed up sth with git and github automatically requested more reviewers. PR is fine now, but I can't undo reviews request.

**Comment by bartoszkubicki** [comment]
Sorry guys, while working on on this PR, I have messed up sth with git and github automatically requested more reviewers. PR is fine now, request reviews is also reverted.

Meanwhile, I have started moving out public methos which should be part of interface, but they legacy from old models. So far I have worked with `getAvailableStatues` from both classes. I have created data providers for both models, and tried to replace all occurances of method, making deprecated all classes using it (by the way using deprecated models). I tried to preserve backward compatibility. For now I have kept method as deprecated in new implementations, but I am not sure if it is necessary. I also created providers objects manually to make method removal easy and possible.

@sivaschenko @sidolov @ishakhsuvarov 
Before moving on I would like to see any review about this action `getAvailableStatues` if it is the right direction and I should do the same with othere methos we are considering to be removed from new implementations. 
If any tests will fail I will try to fix it immediately, so codebase of PR is not currently breaking anything.

**Comment by bartoszkubicki** [comment]
@sivaschenko @sidolov @ishakhsuvarov
There are quite a few things to do in this PR - first of all I need some feedback. 
Although, this PR has been moved to 2.4 release line, maybe we could add breaking changes - they require way less to do. If not, I am waiting for your help.

**Comment by lbajsarowicz** [comment]
@magento run all tests

**Comment by bartoszkubicki** [comment]
Ok, enough. If anyone wants to continue, feel free to pick it up. For Magento 2.4 maybe it is easier way to achieve this - just make backward incompatible changes - I don't know if its ok or not, question is left without answer. I close this PR due to inactivity of magento team.

**Pull request closed without merging**